lib/router/src/cache.rs

```diff
@@ -1,426 +1,322 @@
+    ///
+    /// If a value is returned, this key will not be considered for eviction
+    /// for another `expires` span of time.
+    pub fn access(&mut self, key: &K) -> Option<V> {
+        if let Some(node) = self.values.get_mut(key) {
+            self.expirations.reset(&node.dq_key, self.expires);
+            trace!("reset expiration for cache value associated with key");
+
+            return Some(node.value.clone());
+        }
 
-impl<T> Node<T> {
-    pub fn new(value: T, last_access: Instant) -> Self {
-        Node { value, last_access }
+        None
     }
 
-    pub fn access<'a, N: Now + 'a>(&'a mut self, now: &'a N) -> Access<'a, T, N> {
-        Access { now, node: self }
-    }
+    /// Attempts to insert an item by key.
+    ///
+    /// If a value is returned, this key has been set to expire after an
+    /// `expires` span of time.
+    pub fn insert(&mut self, key: K, value: V) -> Option<V> {
+        let node = {
+            trace!("inserting value into expirations");
+            let dq_key = self.expirations.insert(key.clone(), self.expires);
+            Node { dq_key, value }
+        };
 
-    pub fn last_access(&self) -> Instant {
-        self.last_access
+        trace!("inserting node into values");
+        self.values.insert(key, node).map(|n| n.value)
     }
-}
 
-impl<T> Deref for Node<T> {
-    type Target = T;
-    fn deref(&self) -> &Self::Target {
-        &self.value
-    }
-}
+    /// Evict expired values from the cache.
+    ///
+    /// Polls the underlying `DelayQueue`. When elements are returned from the
+    /// queue, remove the associated key from `values`.
+    fn poll_purge(&mut self) -> Poll<(), Error> {
+        while let Some(expired) = try_ready!(self.expirations.poll()) {
+            trace!("expired an entry from expirations");
+            self.values.remove(expired.get_ref());
+            trace!("removed an entry from values!");
+        }
 
-impl<T> DerefMut for Node<T> {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        &mut self.value
+        Ok(Async::Ready(()))
     }
 }
 
-// ===== impl Now =====
-
-/// Default source of time.
-impl Now for () {
-    fn now(&self) -> Instant {
-        Instant::now()
+// ===== impl PurgeCache =====
+
+impl<K, V> Future for PurgeCache<K, V>
+where
+    K: Clone + Eq + Hash,
+    V: Clone,
+{
+    type Item = ();
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            try_ready!(self.interval.poll().map_err(|e| {
+                panic!("Interval::poll must not fail: {}", e);
+            }));
+            trace!("purge cache interval reached");
+
+            let mut cache = try_ready!(Ok(self.cache.poll_lock()));
+
+            trace!("purge cache is polling for evictions...");
+            match cache.poll_purge() {
+                Ok(Async::NotReady) => return Ok(Async::NotReady),
+                Ok(Async::Ready(())) => (),
+                Err(e) => {
```

Comment:
If you make the error type of `poll_purge` `Never`, you can avoid the panic by matching on Never...

Reply:
I'm inclined to get rid of the match statement since we don't actually need it and change it to `cache.poll_purge().expect("Cache::poll_purge must not fail");`

Reply:
Plus can't `poll_purge` can error if `expirations.poll()` errors?

Reply:
I think it's reasonable to move the `expect` into `poll_purge`, and have it return `Poll<(), Never>`. Not a blocker IMO.

Reply:
This introduces some further complexity into `poll_purge` as the `try_ready!` has to change to a match. I'm going to punt on this for now and consider in the follow-up.